File: videogen-sdk/src/api/resources/tools/tools.rs
use crate::api::*;
use crate::{ApiError, ClientConfig, HttpClient, QueryBuilder, RequestOptions};
use reqwest::Method;

pub struct ToolsClient {
    pub http_client: HttpClient,
}

impl ToolsClient {
    pub fn new(config: ClientConfig) -> Result<Self, ApiError> {
        Ok(Self {
            http_client: HttpClient::new(config.clone())?,
        })
    }

    /// Generate an image from a text prompt, optionally guided by one or more reference images. When reference images are provided, the prompt describes the desired transformation. VideoGen automatically routes each request to the most effective state-of-the-art image model for your prompt, reference images, and quality tier, so you don't pick a model.
    ///
    /// # Arguments
    ///
    /// * `options` - Additional request options such as headers, timeout, etc.
    ///
    /// # Returns
    ///
    /// JSON response from the API
    pub async fn generate_image(
        &self,
        request: &GenerateImageRequest,
        options: Option<RequestOptions>,
    ) -> Result<StartToolExecutionResponse, ApiError> {
        self.http_client
            .execute_request(
                Method::POST,
                "v1/tools/generate-image",
                Some(serde_json::to_value(request).map_err(ApiError::Serialization)?),
                None,
                options,
            )
            .await
    }

    /// Generate a single short video clip (up to 15 seconds) from a text prompt, optionally guided by reference images, videos, and audio. At least one of `prompt`, `imageFileIds`, `videoFileIds`, or `audioFileIds` must be provided. VideoGen automatically routes each request to the most effective state-of-the-art video model for your inputs and settings, so you don't pick a model. This endpoint returns one standalone clip. For longer, higher-quality, professionally edited videos with narration, captions, music, and multiple scenes, use a video workflow such as [Script to video](/workflows) (`POST /v1/workflows/script-to-video`) instead.
    ///
    /// # Arguments
    ///
    /// * `options` - Additional request options such as headers, timeout, etc.
    ///
    /// # Returns
    ///
    /// JSON response from the API
    pub async fn generate_video_clip(
        &self,
        request: &GenerateVideoClipRequest,
        options: Option<RequestOptions>,
    ) -> Result<StartToolExecutionResponse, ApiError> {
        self.http_client
            .execute_request(
                Method::POST,
                "v1/tools/generate-video-clip",
                Some(serde_json::to_value(request).map_err(ApiError::Serialization)?),
                None,
                options,
            )
            .await
    }

    /// Convert text into a spoken audio file. Only voices with `supportsDirectToolExecution` set to true can be used. Optionally choose a voice, language, speed, and pronunciation overrides.
    ///
    /// # Arguments
    ///
    /// * `options` - Additional request options such as headers, timeout, etc.
    ///
    /// # Returns
    ///
    /// JSON response from the API
    pub async fn text_to_speech(
        &self,
        request: &TextToSpeechRequest,
        options: Option<RequestOptions>,
    ) -> Result<StartToolExecutionResponse, ApiError> {
        self.http_client
            .execute_request(
                Method::POST,
                "v1/tools/text-to-speech",
                Some(serde_json::to_value(request).map_err(ApiError::Serialization)?),
                None,
                options,
            )
            .await
    }

    /// Generate a sound effect from a text description. Optionally control the duration and prompt influence. VideoGen automatically routes each request to the most effective state-of-the-art sound effect model for your prompt and settings, so you don't pick a model.
    ///
    /// # Arguments
    ///
    /// * `options` - Additional request options such as headers, timeout, etc.
    ///
    /// # Returns
    ///
    /// JSON response from the API
    pub async fn generate_sound_effect(
        &self,
        request: &GenerateSoundEffectRequest,
        options: Option<RequestOptions>,
    ) -> Result<StartToolExecutionResponse, ApiError> {
        self.http_client
            .execute_request(
                Method::POST,
                "v1/tools/generate-sound-effect",
                Some(serde_json::to_value(request).map_err(ApiError::Serialization)?),
                None,
                options,
            )
            .await
    }

    /// Generate an instrumental music track from a text description. The returned track is approximately 30 seconds long. VideoGen automatically routes each request to the most effective state-of-the-art music model for your prompt, so you don't pick a model.
    ///
    /// # Arguments
    ///
    /// * `options` - Additional request options such as headers, timeout, etc.
    ///
    /// # Returns
    ///
    /// JSON response from the API
    pub async fn generate_music(
        &self,
        request: &GenerateMusicRequest,
        options: Option<RequestOptions>,
    ) -> Result<StartToolExecutionResponse, ApiError> {
        self.http_client
            .execute_request(
                Method::POST,
                "v1/tools/generate-music",
                Some(serde_json::to_value(request).map_err(ApiError::Serialization)?),
                None,
                options,
            )
            .await
    }

    /// Generate a talking-head avatar video by pairing a presenter with an audio file, typically from a prior text-to-speech result.
    ///
    /// # Arguments
    ///
    /// * `options` - Additional request options such as headers, timeout, etc.
    ///
    /// # Returns
    ///
    /// JSON response from the API
    pub async fn generate_avatar(
        &self,
        request: &GenerateAvatarRequest,
        options: Option<RequestOptions>,
    ) -> Result<StartToolExecutionResponse, ApiError> {
        self.http_client
            .execute_request(
                Method::POST,
                "v1/tools/generate-avatar",
                Some(serde_json::to_value(request).map_err(ApiError::Serialization)?),
                None,
                options,
            )
            .await
    }

    /// Convert any raster image into a scalable vector graphic (SVG). The output traces the shapes and colors of the input image.
    ///
    /// # Arguments
    ///
    /// * `options` - Additional request options such as headers, timeout, etc.
    ///
    /// # Returns
    ///
    /// JSON response from the API
    pub async fn vectorize_image(
        &self,
        request: &ImageAssetRequest,
        options: Option<RequestOptions>,
    ) -> Result<StartToolExecutionResponse, ApiError> {
        self.http_client
            .execute_request(
                Method::POST,
                "v1/tools/vectorize-image",
                Some(serde_json::to_value(request).map_err(ApiError::Serialization)?),
                None,
                options,
            )
            .await
    }

    /// Remove the background from an image, returning a transparent-background PNG of the foreground subject.
    ///
    /// # Arguments
    ///
    /// * `options` - Additional request options such as headers, timeout, etc.
    ///
    /// # Returns
    ///
    /// JSON response from the API
    pub async fn remove_image_background(
        &self,
        request: &ImageAssetRequest,
        options: Option<RequestOptions>,
    ) -> Result<StartToolExecutionResponse, ApiError> {
        self.http_client
            .execute_request(
                Method::POST,
                "v1/tools/remove-image-background",
                Some(serde_json::to_value(request).map_err(ApiError::Serialization)?),
                None,
                options,
            )
            .await
    }

    /// Remove the background from a video, producing a transparent-background video of the foreground subject.
    ///
    /// # Arguments
    ///
    /// * `options` - Additional request options such as headers, timeout, etc.
    ///
    /// # Returns
    ///
    /// JSON response from the API
    pub async fn remove_video_background(
        &self,
        request: &VideoAssetRequest,
        options: Option<RequestOptions>,
    ) -> Result<StartToolExecutionResponse, ApiError> {
        self.http_client
            .execute_request(
                Method::POST,
                "v1/tools/remove-video-background",
                Some(serde_json::to_value(request).map_err(ApiError::Serialization)?),
                None,
                options,
            )
            .await
    }

    /// Increase the resolution of an image while preserving detail and sharpness.
    ///
    /// # Arguments
    ///
    /// * `options` - Additional request options such as headers, timeout, etc.
    ///
    /// # Returns
    ///
    /// JSON response from the API
    pub async fn upscale_image(
        &self,
        request: &ImageAssetRequest,
        options: Option<RequestOptions>,
    ) -> Result<StartToolExecutionResponse, ApiError> {
        self.http_client
            .execute_request(
                Method::POST,
                "v1/tools/upscale-image",
                Some(serde_json::to_value(request).map_err(ApiError::Serialization)?),
                None,
                options,
            )
            .await
    }

    /// Increase the resolution of a video while preserving detail and sharpness.
    ///
    /// # Arguments
    ///
    /// * `options` - Additional request options such as headers, timeout, etc.
    ///
    /// # Returns
    ///
    /// JSON response from the API
    pub async fn upscale_video(
        &self,
        request: &VideoAssetRequest,
        options: Option<RequestOptions>,
    ) -> Result<StartToolExecutionResponse, ApiError> {
        self.http_client
            .execute_request(
                Method::POST,
                "v1/tools/upscale-video",
                Some(serde_json::to_value(request).map_err(ApiError::Serialization)?),
                None,
                options,
            )
            .await
    }

    /// Turn a still image into a short video clip with a 3D parallax motion effect, simulating camera movement through the scene.
    ///
    /// # Arguments
    ///
    /// * `options` - Additional request options such as headers, timeout, etc.
    ///
    /// # Returns
    ///
    /// JSON response from the API
    pub async fn image3d_effect(
        &self,
        request: &ImageAssetRequest,
        options: Option<RequestOptions>,
    ) -> Result<StartToolExecutionResponse, ApiError> {
        self.http_client
            .execute_request(
                Method::POST,
                "v1/tools/image-3d-effect",
                Some(serde_json::to_value(request).map_err(ApiError::Serialization)?),
                None,
                options,
            )
            .await
    }

    /// Request cancellation of a running tool execution. The execution transitions to `cancelled` if it has not already completed.
    ///
    /// # Arguments
    ///
    /// * `tool_execution_id` - The tool execution id returned when the tool was started.
    /// * `options` - Additional request options such as headers, timeout, etc.
    ///
    /// # Returns
    ///
    /// JSON response from the API
    pub async fn cancel_tool_execution(
        &self,
        tool_execution_id: &str,
        options: Option<RequestOptions>,
    ) -> Result<StartToolExecutionResponse, ApiError> {
        self.http_client
            .execute_request(
                Method::POST,
                &format!("v1/tools/executions/{}/cancel", tool_execution_id),
                None,
                None,
                options,
            )
            .await
    }

    /// List tool executions started via the API, most recently created first. Use `selfOnly=true` to restrict results to the calling API key's user; otherwise all executions for the team are returned. Cursor-paginated; see the [Pagination](/pagination) guide.
    ///
    /// # Arguments
    ///
    /// * `limit` - Maximum number of items to return in the page. Defaults to 50; capped at 200. See [Pagination](/pagination).
    /// * `cursor` - Opaque pagination cursor returned as `nextCursor` by the previous page. Omit on the first request. Cursors are tied to the endpoint that produced them and must be passed unmodified. See [Pagination](/pagination).
    /// * `self_only` - When true, returns only items created by the API key's owner. When false (default), returns all items accessible to the team.
    /// * `options` - Additional request options such as headers, timeout, etc.
    ///
    /// # Returns
    ///
    /// JSON response from the API
    pub async fn list_tool_executions(
        &self,
        request: &ListToolExecutionsQueryRequest,
        options: Option<RequestOptions>,
    ) -> Result<ToolExecutionListResponse, ApiError> {
        self.http_client
            .execute_request(
                Method::GET,
                "v1/tools/executions",
                None,
                QueryBuilder::new()
                    .int("limit", request.limit.clone())
                    .string("cursor", request.cursor.clone())
                    .bool("selfOnly", request.self_only.clone())
                    .build(),
                options,
            )
            .await
    }

    /// Retrieve the current status and result of a tool execution. Poll this endpoint until `status` is `succeeded`, `failed`, or `cancelled`.
    ///
    /// # Arguments
    ///
    /// * `tool_execution_id` - The tool execution id returned when the tool was started.
    /// * `options` - Additional request options such as headers, timeout, etc.
    ///
    /// # Returns
    ///
    /// JSON response from the API
    pub async fn get_tool_execution_info(
        &self,
        tool_execution_id: &str,
        options: Option<RequestOptions>,
    ) -> Result<ExecutedTool, ApiError> {
        self.http_client
            .execute_request(
                Method::GET,
                &format!("v1/tools/executions/{}", tool_execution_id),
                None,
                None,
                options,
            )
            .await
    }
}

File: src/commands/tools.ts
import type { ParsedArgs } from "../args";
import { resolveRequest } from "../body";
import { printJson } from "../output";
import { paginationFlags, requireStringFlag, selfOnlyFlag } from "./flags";
import type { ResourceCommands } from "./types";

const runTool = async <T extends object>({
  args,
  wait,
  start,
  andWait,
}: {
  args: ParsedArgs;
  wait: boolean;
  start: (request: T) => Promise<unknown>;
  andWait: (request: T) => Promise<unknown>;
}): Promise<void> => {
  // We intentionally use an unsafe `as` assertion here because the API server validates the body and returns a type error if invalid.
  const request = (await resolveRequest({ flags: args.flags, requireBody: true })) as T;
  printJson(wait ? await andWait(request) : await start(request));
};

export const toolsCommands: ResourceCommands = {
  // @sdk-operation generateImage
  "generate-image": async (ctx) =>
    runTool<Parameters<typeof ctx.client.tools.generateImage>[0]>({
      args: ctx.args,
      wait: ctx.wait,
      start: (request) => ctx.client.tools.generateImage(request),
      andWait: (request) => ctx.client.tools.generateImageAndWait(request),
    }),

  // @sdk-operation generateVideoClip
  "generate-video-clip": async (ctx) =>
    runTool<Parameters<typeof ctx.client.tools.generateVideoClip>[0]>({
      args: ctx.args,
      wait: ctx.wait,
      start: (request) => ctx.client.tools.generateVideoClip(request),
      andWait: (request) => ctx.client.tools.generateVideoClipAndWait(request),
    }),

  // @sdk-operation generateMotionGraphic
  "generate-motion-graphic": async (ctx) =>
    runTool<Parameters<typeof ctx.client.tools.generateMotionGraphic>[0]>({
      args: ctx.args,
      wait: ctx.wait,
      start: (request) => ctx.client.tools.generateMotionGraphic(request),
      andWait: (request) => ctx.client.tools.generateMotionGraphicAndWait(request),
    }),

  // @sdk-operation textToSpeech
  "text-to-speech": async (ctx) =>
    runTool<Parameters<typeof ctx.client.tools.textToSpeech>[0]>({
      args: ctx.args,
      wait: ctx.wait,
      start: (request) => ctx.client.tools.textToSpeech(request),
      andWait: (request) => ctx.client.tools.textToSpeechAndWait(request),
    }),

  // @sdk-operation generateSoundEffect
  "generate-sound-effect": async (ctx) =>
    runTool<Parameters<typeof ctx.client.tools.generateSoundEffect>[0]>({
      args: ctx.args,
      wait: ctx.wait,
      start: (request) => ctx.client.tools.generateSoundEffect(request),
      andWait: (request) => ctx.client.tools.generateSoundEffectAndWait(request),
    }),

  // @sdk-operation generateMusic
  "generate-music": async (ctx) =>
    runTool<Parameters<typeof ctx.client.tools.generateMusic>[0]>({
      args: ctx.args,
      wait: ctx.wait,
      start: (request) => ctx.client.tools.generateMusic(request),
      andWait: (request) => ctx.client.tools.generateMusicAndWait(request),
    }),

  // @sdk-operation generateAvatar
  "generate-avatar": async (ctx) =>
    runTool<Parameters<typeof ctx.client.tools.generateAvatar>[0]>({
      args: ctx.args,
      wait: ctx.wait,
      start: (request) => ctx.client.tools.generateAvatar(request),
      andWait: (request) => ctx.client.tools.generateAvatarAndWait(request),
    }),

  // @sdk-operation vectorizeImage
  "vectorize-image": async (ctx) =>
    runTool<Parameters<typeof ctx.client.tools.vectorizeImage>[0]>({
      args: ctx.args,
      wait: ctx.wait,
      start: (request) => ctx.client.tools.vectorizeImage(request),
      andWait: (request) => ctx.client.tools.vectorizeImageAndWait(request),
    }),

  // @sdk-operation removeImageBackground
  "remove-image-background": async (ctx) =>
    runTool<Parameters<typeof ctx.client.tools.removeImageBackground>[0]>({
      args: ctx.args,
      wait: ctx.wait,
      start: (request) => ctx.client.tools.removeImageBackground(request),
      andWait: (request) => ctx.client.tools.removeImageBackgroundAndWait(request),
    }),

  // @sdk-operation removeVideoBackground
  "remove-video-background": async (ctx) =>
    runTool<Parameters<typeof ctx.client.tools.removeVideoBackground>[0]>({
      args: ctx.args,
      wait: ctx.wait,
      start: (request) => ctx.client.tools.removeVideoBackground(request),
      andWait: (request) => ctx.client.tools.removeVideoBackgroundAndWait(request),
    }),

  // @sdk-operation upscaleImage
  "upscale-image": async (ctx) =>
    runTool<Parameters<typeof ctx.client.tools.upscaleImage>[0]>({
      args: ctx.args,
      wait: ctx.wait,
      start: (request) => ctx.client.tools.upscaleImage(request),
      andWait: (request) => ctx.client.tools.upscaleImageAndWait(request),
    }),

  // @sdk-operation upscaleVideo
  "upscale-video": async (ctx) =>
    runTool<Parameters<typeof ctx.client.tools.upscaleVideo>[0]>({
      args: ctx.args,
      wait: ctx.wait,
      start: (request) => ctx.client.tools.upscaleVideo(request),
      andWait: (request) => ctx.client.tools.upscaleVideoAndWait(request),
    }),

  // @sdk-operation image3dEffect
  "image-3d-effect": async (ctx) =>
    runTool<Parameters<typeof ctx.client.tools.image3dEffect>[0]>({
      args: ctx.args,
      wait: ctx.wait,
      start: (request) => ctx.client.tools.image3dEffect(request),
      andWait: (request) => ctx.client.tools.image3dEffectAndWait(request),
    }),

  // @sdk-operation listToolExecutions
  "list-tool-executions": async ({ client, args }) => {
    printJson(
      await client.tools.listToolExecutions({
        ...paginationFlags(args.flags),
        ...selfOnlyFlag(args.flags),
      }),
    );
  },

  // @sdk-operation getToolExecutionInfo
  "get-tool-execution-info": async ({ client, args }) => {
    const toolExecutionId = requireStringFlag(args.flags, "toolExecutionId");
    printJson(await client.tools.getToolExecutionInfo({ toolExecutionId }));
  },

  // @sdk-operation cancelToolExecution
  "cancel-tool-execution": async ({ client, args }) => {
    const toolExecutionId = requireStringFlag(args.flags, "toolExecutionId");
    printJson(await client.tools.cancelToolExecution({ toolExecutionId }));
  },
};
